### PR TITLE
Avoid flushing headers when the server returns a single response

### LINKED ISSUE #4884: Avoid unnecessary flushes for unary responses
We currently optimize flushes for unary requests on client-side, by delaying flushing until the end of the RPC. When looking at the code, I realized [it doesn't appear we're doing that for server-side](https://github.com/grpc/grpc-java/blob/60a0b0c/core/src/main/java/io/grpc/internal/ServerCallImpl.java#L140).

Using wireshare with the interop client/server with empty_unary, we can see a single packet for the request but three packets for the response:
too many server flushes

We should optimize the response flow so that all three frames are sent with a single flush.

### PR DESCRIPTION
**Pull request by amirhadadi**
resolves #4884

### CODE REVIEW

File: core/src/test/java/io/grpc/internal/AbstractTransportTest.java [hunk 1]
```diff
@@ -2042,6 +2042,10 @@ public void clientChecksInboundMetadataSize_trailer() throws Exception {
     assertNull(metadata.get(tellTaleKey));
   }
 
+  private boolean shouldServerFlushHeaders() {
+    return !methodDescriptor.getType().serverSendsOneMessage();
```
Comment (ejona86) [inline core/src/test/java/io/grpc/internal/AbstractTransportTest.java:2046]: The method type used in these tests is UNKNOWN, so they just do whatever they want to, essentially (UNKNOWN works like BIDI_STREAMING, but the transport generally doesn't care too much). That is to say, this is a bit misleading and tests can use other values, so I think we should just hard-code true/false in the tests of this class. (Feel free to do `/*flush=*/ true` or the like it make it more clear, if you'd like.)

**Review by ejona86 — APPROVED** [review]
binder/ also needs updating. You didn't notice it before because `-PskipAndroid=true` will skip it.

File: core/src/test/java/io/grpc/internal/AbstractTransportTest.java (hunk 1 above)
Comment (amirhadadi) [reply core/src/test/java/io/grpc/internal/AbstractTransportTest.java]: Changed to hardcoded flush parameter.

**Comment by ejona86** [comment]
The history has gotten convoluted here and seems broken, and is breaking the diff. Are you fine with me cleaning it up? (Doing so will require force-pushing to the branch, and I'm not entirely sure if I have access to do that but I can certainly try.)

**Review by ejona86 — APPROVED** [review]
I went ahead and cleaned up the history to fix the diff.

**Comment by larry-safran** [comment]
@amirhadadi Would you please complete the EasyCLA and rebase to resolve the conflicts?

**Comment by amirhadadi** [comment]
@larry-safran as we traced the issue of extra segments being sent to flow control [PINGs](https://github.com/grpc/grpc-java/issues/4884#issuecomment-1184553980), not sure it's worth merging this PR.
WDYT? Should it be closed?

**Comment by ejona86** [comment]
With https://github.com/grpc/grpc-java/pull/9847 , we won't be sending PINGs as frequently, especially for small requests/responses on a warmed connection.

Seems this wasn't merged just waiting on @sanjaypujare's review.

**Comment by sanjaypujare** [comment]
> With #9847 , we won't be sending PINGs as frequently, especially for small requests/responses on a warmed connection.
> 
> Seems this wasn't merged just waiting on @sanjaypujare's review.

Yes, I noticed that a couple of days ago. Is there still interest in merging this PR? From yours and @amirhadadi 's comments I am guessing this PR may not be needed anymore. Let me know

**Comment by ejona86** [comment]
My comment was to say this has use, as the thing that made it less useful has been fixed. (And even back when it didn't demonstrate a win, I was approving of it.)

**Comment by sanjaypujare** [comment]
Okay. @amirhadadi do you want to rebase then I will review. 3 files have conflicts

**Review by sanjaypujare — APPROVED** [review]
LGTM

**Pull request merged**